Bus route

alrapal opened the pull request:
resolves #74 

The bus driver gets a route assigned on connection.
They have the bus line as well as the next stop displayed.
They press a button to change to the next stop and send it to passengers via MQTT
By pressing on the current stop, a list with all stops for the line is displayed. It goes away by pressing on the current stop again
When the last stop is reached, the driver and passengers are notified that this is the last stop. The driver is invited to press the "NEXT button again to reverse the line and continue the trip but the other way.

Passenger profile and driver profile are communicating through mqtt the different information needed.
Messages are optimised to be as efficient as possible. 
The information are only sent when the passenger is sending a request when opening the app. 


https://user-images.githubusercontent.com/71485862/119638284-cb0b3780-be16-11eb-8355-9e64c9e4fcfd.mov

alrapal committed:
Bus route
##########################################################
Type:
User story
Tags:
#74
Changes:
- Added a class BusLine.java
- Implemented the different methods regarding the BusLine.java class
- Fixed a bug for the joystick being outside the screen
- Added the interface to display the next stop and the current bus line.
Notes:
Need to fix when the bus has no stops left. Currently crashes
Make the text a bit nicer and complete
Initiate the first stop so it is not empty
Reverse route when done

alrapal committed:
Bus route
##########################################################
Type:
User story
Tags:
#74
Changes:
- Added a test class for BusLine.java
- Reverse correctly when end of line
- Prepared work for displaying all stops
- Fixed gradle so it allows unit testing
Notes:

alrapal committed:
Bus route
##########################################################
Type:
User story
Tags:
resolves #74
Changes:
- When pressing the next stop, a list of all stops for the line is displayed.
- Added animation to the list so it looks nicer.
- Changed back the gradle since it makes the CI bug.
- Deleted duplicated tests
- Update gradle files
Notes:
None

alrapal committed:
Bus route
##########################################################
Type:
User story
Tags:
resolves #74
Changes:
- When pressing the next stop, a list of all stops for the line is displayed.
- Added animation to the list so it looks nicer.
- Changed back the gradle since it makes the CI bug.
Notes:
None

alrapal committed:
Merge remote-tracking branch 'origin/bus_route' into bus_route

# Conflicts:
#	UI/app/build.gradle

alrapal committed:
Merge remote-tracking branch 'origin/master' into bus_route

# Conflicts:
#	UI/app/src/main/java/com/example/pathfinder/Activities/DriverDashboard.java

alrapal committed:
Bus route - bug fix
##########################################################
Type:
User story
Tags:
resolves #74
Changes:
- fixed missing methods from previous merge
Notes:
None

juliavankirk commented on UI/app/src/main/res/layout/activity_driver_dashboard.xml:
```diff
@@ -330,10 +329,98 @@
+                app:layout_constraintHorizontal_bias="0.483"
+                app:layout_constraintStart_toEndOf="@+id/nextStopView"
+                app:layout_constraintTop_toTopOf="parent"
+                app:layout_constraintVertical_bias="0.64" />
```
Don't we already have a button for next stop?

alrapal replied:
Mmmh, we have a layout that does not seem to be a button but more of an indicator of a passenger asking to stop at the next stop

alrapal replied:
But if not, we can do so that this indicator is also a button, which is not a lot of changed in the logic =)

juliavankirk replied:
You can set an onClick listener in the nextStop function I set up. It's been deleted though, so may have to put it back. Just for UX and UI, might be better to have one button for the same function :)

juliavankirk commented on UI/app/src/main/java/com/example/pathfinder/Activities/DriverDashboard.java:
```diff
@@ -69,6 +91,9 @@ protected void onCreate(Bundle savedInstanceState) {
         getWindow().setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN, WindowManager.LayoutParams.FLAG_FULLSCREEN);
         setContentView(R.layout.activity_driver_dashboard);
 
+        bottomAwayAnim = AnimationUtils.loadAnimation(this, R.anim.slide_away_down_animation);
```
I can't seem to simulate the animation on my end. Is this working as intended?

https://user-images.githubusercontent.com/38304351/119358267-bf016780-bca8-11eb-9b2f-7c57f13d8041.mov

alrapal replied:
Works like this locally (with the changes you required, going to push in a minute)
https://user-images.githubusercontent.com/71485862/119360514-fb35c780-bcaa-11eb-8d00-532199942d7c.mov

alrapal replied:
And yeah it is, just did not rename the animation. I thought about sliding away but then I thought it get in the way of the thumbstick. I can rename the animation for more clarity.

juliavankirk replied:
Ah I see, the animation/ list didn't pop up for me even when I pushed the previous "NEXT" button.  Will test again, thanks.

Oh don't change the naming! I was referring to its functionality, since I can see it instantiated in the code but it didn't show anything in the simulator.

juliavankirk commented on UI/app/src/main/java/com/example/pathfinder/Activities/DriverDashboard.java:
```diff
@@ -288,8 +320,77 @@ public void brakeBtn(View view) {
+     * The passengers should subscribe to this topic when choosing the line they are using.
+     * @param comingStop -> the next stop as a string
+     */
+    private void publishNextStop(String comingStop){
```
We should store this information as a shared preference as opposed to sending via MQTT, since it is data that is only client-facing. Unless the data is being passed onto the car itself, we should avoid message transport via MQTT as we already have a tremendous backlog that is increasing our latency.

alrapal replied:
The problem if we use the shared preference, is that we are not sending data from one user to the other but from one profile to the other on the same device. Which is not what it is supposed to do. 
This joins the conversation on the #85, I would like more insights from the rest of the group regarding that. 

On the other hand, this is for a user story that is not implemented and is low priority #75. We should have a meeting and decide if it is worth implementing it. And if so, if we use MQTT or shared preference. 

I understand the latency issue, I am just not sure what to decide about the shared preference solution. And I am not able to test the MQTT because of the camera feed which is killing my computer.

juliavankirk replied:
Pretend you're a bored passenger using the Pathfinder app, and you wanna do something else while you wait. You click out of the app and go on Spotify, Facebook, whatever. Shared preference saves your last input, and loads your last input from when you last ran the app.

So, if you pressed the stop button, and somehow left the app, when you come back it will still have the Stop button toggled on. Without it, even if you saved the message into MQTT, the sign may flash but the stop and accessibility buttons would default back to off. This would be very confusing to the user. 
If you have tinkered with the passenger_screen branch, you'd see that the "Stop" button changes text to "Cancel Stop" and vice versa when toggled.

An idea that comes to mind would be to publish an MQTT message in a shared preference function. That way all necessary data gets stored and you get to pass data along from one process to another.

alrapal committed:
Bus route - bug fix
##########################################################
Type:
User story
Tags:
resolves #74
Changes:
- Change button following @juliavankirk review.
- The "NEXT" button displays the list for more clarity (could be confusing to press the displayed next stop)
Notes:
None

alrapal committed:
Bus route - bug fix
##########################################################
Type:
User story
Tags:
resolves #74
Changes:
- Change button following @juliavankirk review.
- The "NEXT" button displays the list for more clarity (could be confusing to press the displayed next stop)
- Added some comments for methods
- Changed some naming to match the effect
Notes:
None

bassamEldesouki commented:
Hey Alex, I am concerned about that the "Stop list" is covering other functionalities when clicking it. I know the layout is already stuffed, maybe we could have another layout shows these info before proceeding to the driver's control dashboard? And then the dashboard could just show the next stop. Just a suggestion! We can  also discuss that in the meeting tomorrow :)

alrapal commented:
This PR is closed and will come back with new elements. The original issue #88  has been completed with new elements which make this PR no complete anymore 

Be back soon ! 😄

alrapal commented:
> Hey Alex, I am concerned about that the "Stop list" is covering other functionalities when clicking it. I know the layout is already stuffed, maybe we could have another layout shows these info before proceeding to the driver's control dashboard? And then the dashboard could just show the next stop. Just a suggestion! We can also discuss that in the meeting tomorrow :)

Hi Bassam ! Yes, I understand. That is how I thought about it at the beginning, but then I thought what if the driver wants to see the list while he is at a stop for instance. It would need to go to the page for it which would make the whole interface hidden (if we think the list is on another activity). 

We can definitely discuss the alternatives during today's meeting, yes =)

alrapal committed:
Bus route - enhancement
##########################################################
Type:
User story
Tags:
#74
Changes:
- Change of style for the list of stops
- Cleaning of code
- Separation of elements into different XLM for more readability.
Notes:
None

alrapal committed:
Bus route - passenger side
##########################################################
Type:
User story
Tags:
#74
Changes:
- Stops are sent from the driver to the passenger via MQTT
Notes:
Need to implement the list for all stops now and the bus line.

alrapal committed:
Bus route - passenger side
##########################################################
Type:
User story
Tags:
#74
Changes:
- Bus name and list of stops is sent from the driver to the passenger via MQTT
- The passenger triggers the message from the driver by sending a number (to limit the size of messages)
- Driver dashboard reacts to triggers and send requested data
- PassengerDashboard parse messages and display information.
Notes:
Work on smoother design need to be done.

alrapal committed:
Bus route - design
##########################################################
Type:
User story
Tags:
resolves #74
Changes:
- Adjustment on the size of text and borders to look nicer.
Notes:

alrapal committed:
Merge remote-tracking branch 'origin/master' into bus_route

# Conflicts:
#	UI/app/src/main/java/com/example/pathfinder/Activities/DriverDashboard.java
#	UI/app/src/main/java/com/example/pathfinder/Activities/PassengerDashboard.java
#	UI/app/src/main/res/layout/activity_driver_dashboard.xml
#	UI/app/src/main/res/layout/activity_passenger_dashboard.xml
#	UI/app/src/main/res/values/strings.xml

ediz-genc commented on UI/app/src/main/java/com/example/pathfinder/Activities/DriverDashboard.java:
```diff
@@ -124,28 +166,32 @@ public void onFailure(IMqttToken asyncActionToken, Throwable exception) {
         //We need the negative of seekBar.getProgress()
-        int seekProgress = - seekBar.getProgress();
-        if(isCruiseControl){
+        int seekProgress = -seekBar.getProgress();
```
Here line 179 is a conflict with master because of (negative as it is) seekbar instance which is named in master before additional "m" in the front. Easy fix ;)

ediz-genc commented on UI/app/src/main/java/com/example/pathfinder/Activities/DriverDashboard.java:
```diff
@@ -156,9 +202,11 @@ public void onCruiseControlBtn(View view) {
+     * If cruise control is enabled, vehicle will drive with speed based on seekBar's progress
+     */
+    private void seekBarListener() {
         seekBar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
```
Here is another one line 210, seekbar instance declared in master as "m" in the front. Easy fix as well.

alrapal replied:
Thanks Ediz!
I think now it should be fine =)

ediz-genc reviewed (changes requested):
Apart from the conflicts with master in "driverdashhboard.java" you might wanna fix before the merge and considering the demo video (25th of May at 22:56) you provided in edited PR, there is a good work there. :)

Note: Unless you don't have anything to add to this straightforward friendly feature.

alrapal committed:
Merge remote-tracking branch 'origin/master' into bus_route

# Conflicts:
#	UI/app/src/main/java/com/example/pathfinder/Activities/DriverDashboard.java

iachella commented on UI/app/src/main/java/com/example/pathfinder/Activities/PassengerDashboard.java:
```diff
@@ -127,9 +167,153 @@ public void onClick(View v) {
+        Log.d("Bus Stop", message_);
+        char[] charList = message_.toCharArray();
+        String stop = "";
+        for (char c : charList){
```
Here you could use the method String.split() on something like
String[]  messageArray = message_.split(";");
To parse the message to a string array? So you won't need the for loop, I think :)

alrapal replied:
Good idea haha ! Why didn't I thought about that, so much simpler xD

alrapal replied:
Ok, I know why haha, I am using a ArrayList for the StopList adapter. and split returns an array, not an array list. So I have to convert the array into an array list afterwards or creating a new/adapt the AdapterClass to use the array, which was more work than parsing "manually"

iachella commented on UI/app/src/main/java/com/example/pathfinder/Model/BusLine.java:
```diff
@@ -0,0 +1,107 @@
+
+    // Attributes
+    private String name;
+    private Stack <String> comingStops;
```
Stacks, nice idea!

iachella reviewed (commented):
Looks good! Good job with the testing also.
I just left a question with a little suggestion :)

juliavankirk reviewed (commented):
Maybe just change the video to reflect the updated screens? :)

ChrisAxt reviewed (changes requested):
Nice work! When you get to the "end of line" should it not show which stop that is? It's not really necessary but would be something the user would appreciate, maybe just something like:  EOL : "stop name"

alrapal commented:
> Nice work! When you get to the "end of line" should it not show which stop that is? It's not really necessary but would be something the user would appreciate, maybe just something like: EOL : "stop name"

Mmmh, maybe that part needs to be more clear. The end of line means that there are no more stops. So maybe I could change that for "Reversing line" or something else ?. Or any other suggestions that might be better ?

alrapal committed:
Change documentation for more accuracy.
Change the end of line mention to be clearer
Passengers know when the next stop is the terminus

alrapal committed:
Adapted tests to new functioning

alrapal committed:
Adapted tests to new functioning

alrapal committed:
Merge remote-tracking branch 'origin/bus_route' into bus_route

# Conflicts:
#	UI/app/src/test/java/com/example/pathfinder/Model/BusLineTest.java

alrapal commented:
@juliavankirk Video updated !
@edizgenc2021 The master is merge correctly now!
@gusaxtcha Fixed the code so it is clearer with the "end of line" text !
😄

alrapal committed:
Passengers get the current next stop directly without needing to wait for the driver to press the button "next stop"
